### PR TITLE
Fixes #25335 - Creating a React/Redux Editor

### PR DESCRIPTION
**Pull request by glekner**
[Discourse](https://community.theforeman.org/t/creating-a-new-redux-react-editor/11627) to discuss new UX

_TODO_

- [x] Preview Action

- [x] Tests

### CODE REVIEW

File: webpack/assets/javascripts/react_app/components/Editor/Editor.js
Comment (amirfefer) [inline webpack/assets/javascripts/react_app/components/Editor/Editor.js]: wouldn't it be better to use some sort of initialization action?

File: webpack/assets/javascripts/react_app/components/Editor/EditorReducer.js
Comment (amirfefer) [inline webpack/assets/javascripts/react_app/components/Editor/EditorReducer.js]: how about moving this logic into the action (why firing an action if it doesn't change the state?) , plus throw an error if it doesn't exist.

File: webpack/assets/javascripts/react_app/components/Editor/Editor.js
Comment (glekner) [reply webpack/assets/javascripts/react_app/components/Editor/Editor.js]: Great suggestion

File: webpack/assets/javascripts/react_app/components/Editor/EditorReducer.js
Comment (glekner) [reply webpack/assets/javascripts/react_app/components/Editor/EditorReducer.js]: Great suggestion 2#

**Comment by glekner** [comment]
@theforeman/ui-ux 
I removed WIP label, this PR is getting mature and the review process should start.

Feedback I need:
- Test the Preview function (is it working properly?)
- What code can be deprecated/removed ? 

Editor views you can try:
- Monitor => Report Templates
- Host => Partition, Provisioning Templates
- Configure => Global Parameters

File: app/views/common_parameters/_form.html.erb
Comment (ohadlevy) [inline app/views/common_parameters/_form.html.erb]: would this work if you have more than one parameter? I believe each div needs to have a unique id.

File: app/helpers/editor_helper.rb
Comment (ohadlevy) [inline app/helpers/editor_helper.rb]: I believe this can be extracted from the object itself,  perhaps using https://api.rubyonrails.org/classes/ActionView/Helpers/FormBuilder.html#method-i-fields_for ?

File: test/integration/report_template_js_test.rb
Comment (ohadlevy) [inline test/integration/report_template_js_test.rb]: why delete all of the tests? I would assume they are still valid?

Comment (glekner) [reply test/integration/report_template_js_test.rb]: The deleted tests are trying to change the editor value, but its not rendered in the html.
I tried putting `wait_for_ajax` but still the value of the `@response` is:

`//<![CDATA[
$(tfm.reactMounter.mount('Editor', '#react-ace', { ...`

If theres a way to fix it then those tests can be fixed.
Anyway, i'm testing the editor in the JS tests.

File: app/views/common_parameters/_form.html.erb
Comment (glekner) [reply app/views/common_parameters/_form.html.erb]: this form is being rendered when you click on one parameter, I have multiple parameters and it works fine.

Comment (ohadlevy) [reply app/views/common_parameters/_form.html.erb]: thanks, I misunderstood this to be the domain parameters (or other parameters) where we dynamically add forms to the UI, can you validate those are not using the same code path?

File: package.json
Comment (ohadlevy) [inline package.json]: [EDIT: PBKAC] packaging PR at https://github.com/theforeman/foreman-packaging/pull/3207

File: test/integration/common_parameter_test.rb
Comment (glekner) [inline test/integration/common_parameter_test.rb]: this functionality is being tested in the JS tests

**Comment by glekner** [comment]
Fixed rb integration tests

**Comment by glekner** [comment]
Deleted all JQuery from `foreman_editor.js` 🎉

File: test/integration/parameters_permissions_domain_test.rb
Comment (ohadlevy) [inline test/integration/parameters_permissions_domain_test.rb]: these are good tests to keep...?

File: test/integration/parameters_permissions_global_js_test.rb
Comment (ohadlevy) [inline test/integration/parameters_permissions_global_js_test.rb]: i assume wait_for_ajax doesnt work? /cc @amirfefer

File: webpack/assets/javascripts/react_app/components/Editor/Editor.fixtures.js
Comment (ohadlevy) [inline webpack/assets/javascripts/react_app/components/Editor/Editor.fixtures.js:33]: why is this defined twice? just wondering why we need it in the rails part to begin with?

Comment (glekner) [reply webpack/assets/javascripts/react_app/components/Editor/Editor.fixtures.js:33]: these are just the fixtures for testing. But the real data do come from the rails, do we prefer it in the client?

**Comment by glekner** [comment]
@ares Can you try the Editor? would like feedback :)

**Comment by ohadlevy** [comment]
@glekner can you rebase please?
@ares I would really love to get this merged soon, will you have time to test out the implementation ?

File: webpack/assets/javascripts/react_app/components/Editor/Editor.fixtures.js
Comment (ohadlevy) [reply webpack/assets/javascripts/react_app/components/Editor/Editor.fixtures.js:33]: if its static list, then yes?

**Comment by glekner** [comment]
@ohadlevy I rebased yesterday, this is ready for testing. Travis w/node 6 fails here like in other PRs I saw.

File: webpack/assets/javascripts/react_app/components/Editor/Editor.js
Comment (sharvit) [inline webpack/assets/javascripts/react_app/components/Editor/Editor.js]: Could you send only the relevant data? It feels risky and we might lose control over what actually sent to this method.

File: webpack/assets/javascripts/react_app/components/Editor/EditorActions.js
Comment (sharvit) [inline webpack/assets/javascripts/react_app/components/Editor/EditorActions.js]: What's the difference between:

```js
initialState.hosts = {
  ...hosts.map(element => ({ [element.host.name]: element.host.id }))
};
```

File: webpack/assets/javascripts/react_app/components/Editor/EditorActions.js
Comment (sharvit) [inline webpack/assets/javascripts/react_app/components/Editor/EditorActions.js]: Can simply be: `payload: initialState`

File: webpack/assets/javascripts/react_app/components/Editor/EditorActions.js
Comment (sharvit) [inline webpack/assets/javascripts/react_app/components/Editor/EditorActions.js]: I am pretty much against the idea of using generic actions and reducers as it goes against a lot of the principles of Redux.

See:
https://stackoverflow.com/questions/48308444/i-created-a-generic-reducer-for-the-state-when-should-it-be-used-and-when-shou
https://softwareengineering.stackexchange.com/questions/322222/creating-generic-actions-in-redux

File: webpack/assets/javascripts/react_app/components/Editor/EditorActions.js
Comment (sharvit) [reply webpack/assets/javascripts/react_app/components/Editor/EditorActions.js]: It's a pattern in all your actions, it's kind of nitpicking, but I see it as redundant.

File: webpack/assets/javascripts/react_app/components/Editor/EditorActions.js
Comment (sharvit) [inline webpack/assets/javascripts/react_app/components/Editor/EditorActions.js:69]: You don't actually need the `dispatch`, can just return this object like in simple actions (it's not async action).

File: webpack/assets/javascripts/react_app/components/Editor/EditorActions.js
Comment (sharvit) [inline webpack/assets/javascripts/react_app/components/Editor/EditorActions.js:91]: `dispatch` is redundant.

File: webpack/assets/javascripts/react_app/components/Editor/EditorReducer.js
Comment (sharvit) [inline webpack/assets/javascripts/react_app/components/Editor/EditorReducer.js]: Can remove the `{` and the `}` and remove 6 lines

File: webpack/assets/javascripts/react_app/components/Editor/components/EditorOptions.js
Comment (sharvit) [inline webpack/assets/javascripts/react_app/components/Editor/components/EditorOptions.js]: Can't it be a component that lives in its own file?

**Review by sharvit — CHANGES_REQUESTED** [review]
Awesome work @glekner :1st_place_medal:

File: webpack/assets/javascripts/react_app/components/Editor/Editor.js
Comment (glekner) [reply webpack/assets/javascripts/react_app/components/Editor/Editor.js]: can i access the state from the action itself with a selector? is it a good idea?

File: webpack/assets/javascripts/react_app/components/Editor/EditorActions.js
Comment (glekner) [reply webpack/assets/javascripts/react_app/components/Editor/EditorActions.js]: This won't work, as it returns an object of objects e.g.
```
 {
   0: {aaron-ellert.ramon-eatherly.gail-walema.example.org: 22}
   1: {abel-bearman.ruben-montieth.example.net: 66}
   2: {alvin-bocchini.erika-kirton.elton-lucore.example.org: 44}
 }
```

my code returns:

```
 {
   aaron-ellert.ramon-eatherly.gail-walema.example.org: 22
   abel-bearman.ruben-montieth.example.net: 66
   alvin-bocchini.erika-kirton.elton-lucore.example.org: 44
 }
```

File: webpack/assets/javascripts/react_app/components/Editor/Editor.js
Comment (sharvit) [reply webpack/assets/javascripts/react_app/components/Editor/Editor.js]: You can access the state from within the action and pass it to a selector (which is one of the reason I try to push us back to selectors that recive the whole state object).

Some developers says you never actually need to do that because you can recive the state into your component props (with selectors) so you can just pass them to your actions.
I bealive in some cases it is more beneficial to do what you described and use the `getState` method.

In this case, do you think it would be more beneficial to get your data directly from the state? Isn't all the data you need already exists for you in the component props?

**Comment by glekner** [comment]
- Removed the generic action, and implemented most of @sharvit suggestions.

**Comment by ohadlevy** [comment]
@glekner can you rebase please? recent changes to parameters imho have changed the editor implementation for global parameters.

@ares it would be great if you can review this asap, so further painful rebases can be avoided...

File: test/integration/parameters_permissions_global_js_test.rb
Comment (ohadlevy) [reply test/integration/parameters_permissions_global_js_test.rb]: @amirfefer can you have a look please?

**Comment by glekner** [comment]
Rebased. Looking at the `_form` changes of the parameters, looks like there is no need for an editor in global parameters anymore, so I removed it.

**Comment by ohadlevy** [comment]
I wonder if it was removed because of the lack of ability to syntax
highlight or some other reason? @ares?

**Comment by ares** [comment]
IIRC it didn't support the way we hide parameters that are set for hidden values. We're using font that changes all characters with dots. Therefore if global parameter contains a password and is marked as "hidden value", by opening the edit form, it does not show it to people looking over a shoulder.

Global parameters can contain erb and now since they can have complex type, the full editor can be useful again. Does it work with hidden values?

**Comment by glekner** [comment]
@ares the new editor supports hiding a value, I added this feature specifically for this view. But it doesn't hide it on default, so for example if a user refreshes the page it will be visible, if we want to save the hidden state I guess we should save it in the db. so do we want to use the new editor there?

**Comment by ares** [comment]
> if we want to save the hidden state I guess we should save it in the db. so do we want to use the new editor there?

I don't fully understand, the attribute is stored in DB so the editor should look at `$parameter.hidden_value` and hide or show the value based on this. I think having advanced editor for parameters would be even more important now when we have advanced types such as array or hash for both parameters and ansible variables. Note that params also can contain erb syntax.

But at the same time, most parameter values are one-line values. So the full editor should probably only open in fullscreen. Anyway, that can be separate PR.

I'm starting testing and will focus on templates editing only now.

**Review by ares — CHANGES_REQUESTED** [review]
First of all, this looks as a great addition. I found few things from user point of view:

1) clicking preview, it does not render the preview until I select some host, it used to render it for the first host in the selector, which is a bit nicer as usually I'm not interested in specific host

2) if I go back to editing, change something and click preview, I see old preview, it's not clear I need to choose the preview host again to see fresh preview

3) the previous selector for host was nicer if I need to switch from one host to another, now I need to click the cross to clean the field and then click again to input to open the list

4) is the list paginated somehow? it looks it just takes preview_host_collection for a given class, that can be 10 000 hosts or more, this is why we had ajax to load matching hosts

5) ~changing SYNTAX in settings does not do anything, it remains as Text~, now I realized it changes to Text when I'm in preview, I can't select anything else then but when I switch back to edit, it changes to the value I selected in preview, which is slightly confusing, though probably acceptable if there's no good solution for this

6) labels in Settings are all capitals, should it be that way?

7) after opening full screen, there seems to be a white bar in the modal at bottom, is this intended?

8) rendering a preview can take some time, there's no indication something is loading

9) there's a white gap between the editor menu and editor text area, it used to be there previously too but not the menu feels more like tabs, perhaps it would look better without the white space?

10) import silently sets the content of editor while I'm in preview mode, maybe we should disable it in preview and change?

11) import file seems to only work once, second time it fails silently

One question - is it possible to specify format when I mount the component? E.g. for ansible playbooks, it would be great to set YAML right away. Not needed in this PR though :-) 

In particular, I like the fact it works with REX plugin out of the box. I like the new diff and also the new design of editor header. Good job. If you're interested I have couple of ideas how to improve the template editing further, but that's outside of scope for this PR :-)

File: app/helpers/editor_helper.rb
Comment (ares) [inline app/helpers/editor_helper.rb]: how do we handle templates introduced by plugins (e.g. JobTemplate form REX plugin), I suppose this is used for form element name, could perhaps this work?
```
"#{template.class.underscore}[template]"
```

though I think "template" attribute will not always the name of attribute for which we mount the editor

**Comment by glekner** [comment]
_Updates_ 🚀 
* Replaced TypeAhead with Select component for the host picker for preview.
* Preview is triggered automatically with the first available host upon going to the preview tab
* Settings labels are not uppercased anymore
* Added a spinner when a preview is loading
* Removed gap between the editor header and the content
* Disabled the Import button when not in the Editor view.

@ares regarding your review:

7. It depends on the longest line in the file, if its wider than the window you will see a scrollbar which is probably white.
10. Import works twice for me, can you explain how to reproduce this bug?

_TODO:_
1) Fetch hosts w/API
2) Warn the user the preview is outdated if he changed the template and navigated to the Preview tab again.

**Comment by glekner** [comment]
_Updates_ 🚀 
* Added a Warning whenever a rendered preview thats displayed is outdated.
* Disabled the options to change Syntax/Keybind on the Preview tab.
* Clearing the host selection will also clear the preview

@ohadlevy @ares  Do you have maybe an advice on how to fetch hosts w/API
that are returned from [this function](https://github.com/theforeman/foreman/blob/develop/app/controllers/hosts_controller.rb#L623)
I guess what is needed is similar to what [we did](https://github.com/theforeman/foreman/pull/6469/files#diff-608d83c9bbc71d670e46c1a9cd1812eaR7) in the WIP audits PR, But ,maybe adding a `format.json` with the preview function to the host controller is too specific because its for templates only?

**Comment by glekner** [comment]
- Added selectors
- Regarding the Select component, it supports an `ajax` func but it can't be attached to a `<select>` which is what the component use. this issue is fixed with Select v4 so I guess i can create another PR which upgrades it or I can even write a new Select component that doesn't use JQuery, anyway its a shame it blocks this PR but I guess its necessary.

**Comment by ares** [comment]
this needs rebase now

**Comment by glekner** [comment]
@ares this was not the branch I meant, but since i finished my work on it I merged it with this PR.
It's now updated, containing the files for the new Select component, which will be deleted next week, since the review is complete on the [PF side.](https://github.com/patternfly/patternfly-react/pull/1650)

This is now updated for you to test.

**Review by ares — CHANGES_REQUESTED** [review]
Thanks for update @glekner 

I gave it another round, here are my findings:

1) pressing enter in editor does not add a new line

2) the first line seems to be missing some top padding
editor_1

3) the new selector now works nicely as expected, can it be styled closer to select2? or could we modify select2 styles so that they look as similar as possible?

Functionality wise, it works really well now! I suppose we'll need to wait until pf-react PR gets in and is released and then we'll need to bump the version?

**Comment by glekner** [comment]
thanks @ares 

1. fixed
2. You said we should remove the white space between the navbar and the editor. we can either bring it back but smaller, or just stay as it is?
3. awsome! about styling i guess we can change it later, and migrate between the two components. I styled it based on the PF design

**Comment by ares** [comment]
For 2. it's good that there's no space between editor and navbar. I meanT padding inside the editor window. It feels like the line height of the first line is too small. I think in css language it's padding-top of the black editor text area that needs to be bigger.

**Comment by glekner** [comment]
- removed `react-click-outside` package and implemented it myself in `EditorHostSelect`
- added padding-top for the first line of the editor

Now waiting for `patternfly-react-extensions` to release a new version, the Select component is already merged. @ares

**Comment by glekner** [comment]
- rebased + using Select component from patternfly

I believe we are ready 🚀

**Comment by glekner** [comment]
[test upgrade]
[test katello]

**Review by sharvit — DISMISSED** [review]
Thanks for the long effort @glekner LGTM :1st_place_medal:

**Comment by mmoll** [comment]
are there RPM PRs needed for this?

**Comment by ohadlevy** [comment]
> are there RPM PRs needed for this?

I did try a long time ago at https://github.com/theforeman/foreman-packaging, probably needs to be revisited...

**Comment by ares** [comment]
There are few more things that don't work
1) go to existing template, change the content then switch to preview, click on revert button and confirm, the editor activates `Editor tab` again but it seems preview is still there, I need to click Preview tab and back to Editor to get the new text content

2) when I click on preview after I updated the text in Editor, I see that it's out of date, that's nice but previously it recreated preview right-away this forces me to do 2 clicks, what's the reason for it? I like the styling of message though :-)

Otherwise works good. We'll need @theforeman/packaging ack though for new dependencies I guess. @glekner mind to try packaging new dependencies?

**Comment by glekner** [comment]
nice finds @ares 

1. fixed
2. I guess I thought about scenarios where you want to go back and forth between the first preview and your code before updating the preview? maybe i'm wrong.

about packaging, i'll try :)

File: package.json
Comment (glekner) [inline package.json:100]: https://github.com/theforeman/foreman-packaging/pull/3706

File: package.json
Comment (glekner) [inline package.json]: https://github.com/theforeman/foreman-packaging/pull/3707

File: app/assets/stylesheets/base.scss
Comment (ohadlevy) [inline app/assets/stylesheets/base.scss:199]: there are some leftover css classes such as .fullscreen* and similar function names in JS.

File: package.json
Comment (ohadlevy) [inline package.json:75]: is it required? I think its actually now bundled in react-ace and we can remove it from here?

File: package.json
Comment (ohadlevy) [inline package.json]: I think you need to package this one too? at the time of writing its 2.32.4  ? (not seeing 2.33?)

File: webpack/assets/javascripts/react_app/components/Editor/EditorActions.js
Comment (ohadlevy) [reply webpack/assets/javascripts/react_app/components/Editor/EditorActions.js:69]: @glekner bump on this ?

File: webpack/assets/javascripts/react_app/components/Editor/Editor.fixtures.js
Comment (ohadlevy) [reply webpack/assets/javascripts/react_app/components/Editor/Editor.fixtures.js:33]: import the fixtures from the const file? (instead of defining twice?)

**Comment by ohadlevy** [comment]
@ares there are minor code nitpicks that needs to be handled, but can you ack from a usability POV ? thanks!

**Comment by ares** [comment]
I'll retest once code is final, there were regressions in last patches, but I don't expect any other findings. I see https://github.com/theforeman/foreman-packaging/pull/3707 has been filed :+1: 

@glekner please ping me when you push the new version, thanks!

**Comment by glekner** [comment]
- removed unused css, reset brace version.

Packaging:
- PF-React - https://github.com/theforeman/foreman-packaging/pull/3752
- PF-Extensions - https://github.com/theforeman/foreman-packaging/pull/3707
- ReactAce - https://github.com/theforeman/foreman-packaging/pull/3706

ping @ares @ohadlevy

File: package.json
Comment (glekner) [inline package.json]: https://github.com/theforeman/foreman-packaging/pull/3752

File: package.json
Comment (ohadlevy) [reply package.json:100]: i still think you need to remove the brace from this file, it is no longer imported by foreman (rather via by react-ace)

Comment (glekner) [reply package.json:100]: @ohadlevy  we need it in `EditorView` - https://github.com/theforeman/foreman/pull/6180/files#diff-43445d906144300b418f2f9a8c540820R5

as stated in Ace docs - https://github.com/securingsincity/react-ace/blob/master/docs/Ace.md

Comment (ohadlevy) [reply package.json:100]: does it make sense to export brace from react-ace? i think its wrong to require it twice (here and at brace) perhaps send them a pr to export it? then we wont have two copies of it...?

Comment (ohadlevy) [reply package.json:100]: @glekner https://github.com/securingsincity/react-ace/issues/638 ?

File: package.json
Comment (ares) [reply package.json]: changes requested there, waiting on packaging ack here, I retested and functionality wise it's ready :+1:

**Review by ares — DISMISSED** [review]
Works nicely now, waiting just on packaging ack

**Review by mmoll — DISMISSED** [review]
Approving, @glekner please look out for comments in the RPM PRs. Leaving the merge up to @ohadlevy.

File: package.json
Comment (ohadlevy) [reply package.json:100]: ok, can you align the brace version again (so they dont conflict with react-ace?) i think this should be enough to unblock this PR. I already created a packaging PR at https://github.com/theforeman/foreman-packaging/pull/3706

Comment (glekner) [reply package.json:100]: aligned

**Comment by mmoll** [comment]
@glekner one last rebase please :)

**Review by ohadlevy — DISMISSED** [review]

**Comment by glekner** [comment]
rebased

File: app/assets/stylesheets/base.scss
Comment (ohadlevy) [reply app/assets/stylesheets/base.scss:199]: @glekner do you want to follow up with another PR after this one gets merge?

**Review by ohadlevy — DISMISSED** [review]

**Comment by ohadlevy** [comment]
[test foreman]

File: package.json
Comment (ohadlevy) [inline package.json:96]: seems like you are missing patternfly-react-extensions after the rebase?

**Review by mmoll — APPROVED** [review]
RPM PRs are open

**Comment by mmoll** [comment]
merged, thanks @glekner!

**Comment by tbrisker** [comment]
RPM building is now failing, the packaging prs weren't merged and i'm now confused which one of them should be merged.

**Pull request merged**
